### PR TITLE
Fix whole editor paste with locked code

### PR DESCRIPTION
**Pull request by KaiSaba**

### CODE REVIEW

File: src/tools.ts [hunk 1]
```diff
@@ -28,6 +28,21 @@ function getRangesFromDecorations (
...
+
+  if (!withDecoration) {
+    return ranges[ranges.length - 1]
+  }
+
+  const firstUneditableRange = ranges[0]
+  return firstUneditableRange != null
+    ? new monaco.Range(fullModelRange.startLineNumber, fullModelRange.startColumn, firstUneditableRange.startLineNumber, firstUneditableRange.startColumn)
+    : undefined
+}
```
Comment (CGNonofr) [inline src/tools.ts:44]: So we only support a single range now? what is the point of computing multiple one then?

We only have some code to invert the ranges in this file around line 320

Comment (KaiSaba) [reply src/tools.ts:44]: I added a generic function that recomputes the editor operations: f538ee0.

Comment (CGNonofr) [reply src/tools.ts:44]: Can the code in the paste callback be removed then?

Comment (KaiSaba) [reply src/tools.ts:44]: Yes, we should be able to remove the modification of `editor.trigger`. I'll removing it and test it.

Comment (KaiSaba) [reply src/tools.ts:134]: Here is the commit that cleans up the code: 04bfcfb.

Comment (KaiSaba) [reply src/tools.ts:135]: I found a bug while testing, here is the fix: c40f3c1.

File: src/tools.ts [hunk 2]
```diff
@@ -28,6 +17,123 @@ function getRangesFromDecorations (
     .map((decoration) => decoration.range)
 }
 
+function minusRanges (uniqueRange: monaco.Range, ranges: monaco.Range[]): monaco.Range[] {
+  const newRanges: monaco.Range[] = []
+  let lastEndLineNumber = uniqueRange.startLineNumber
+  let lastEndColumn = uniqueRange.startColumn
+
+  for (const range of ranges) {
+    const newRange = new monaco.Range(lastEndLineNumber, lastEndColumn, range.startLineNumber, range.startColumn)
```
Comment (CGNonofr) [inline src/tools.ts:26]: nit: you can use a monaco.Position instead of the lineNumber and column

File: src/tools.ts [hunk 3]
```diff
@@ -28,6 +17,123 @@ function getRangesFromDecorations (
...
+    // The operation range doesn't intersect with an uneditable range
+    if (uneditableRangesThatIntersects.length <= 0) {
+      newOperations.push(operation)
+      continue
+    }
+
+    // The operation range is entirely in an uneditable range
+    if (uneditableRangesThatIntersects.some(range => range.containsRange(operationRange))) {
+      displayLockedCodeError(
+        new monaco.Position(operationRange.startLineNumber, operationRange.startColumn))
```
Comment (CGNonofr) [inline src/tools.ts:91]: It's probably not the responsability of this fonction to display an error

File: src/tools.ts (hunk 1 above)
Comment (CGNonofr) [reply src/tools.ts:135]: Can you please split the code a little bit more? I'm struggling to understand the logic

File: src/tools.ts (hunk 3 above)
Comment (KaiSaba) [reply src/tools.ts]: I removed the call to `displayLockedCodeError` while refactoring the function: b598071.

File: src/tools.ts (hunk 1 above)
Comment (KaiSaba) [reply src/tools.ts:178]: Here is the refactor:  b598071.

I don't think I can split more the functions.

File: src/tools.ts (hunk 2 above)
Comment (KaiSaba) [reply src/tools.ts]: 7ee4792

File: src/tools.ts (hunk 1 above)
Comment (KaiSaba) [reply src/tools.ts]: I moved the functions in other files: 8faaeb6.

I changed the logic to split the ranges, then split the operation text separately: c291e16.

Comment (KaiSaba) [reply src/tools.ts]: Here are the commits that fix the last bugs: 8412173 and e47df09.

**Comment by KaiSaba** [comment]
@CGNonofr I added 3 commits to refactor and fix the remaining bugs:
- f8cea32: refactor: refactor minusRanges function to return all ranges
- f8c1ed4: fix: use minusRanges to get hidden areas and visible ranges in hideCodeWithoutDecoration
- 0c5ae90: fix: fix split operation text function when uneditable range text not in text to split

**Comment by KaiSaba** [comment]
@CGNonofr I added a commit to fix the line added on paste:
- 8c23db2: fix: fix locked code ranges and text split

File: src/tests/utils.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+export function createTestRange (
+  model: monaco.editor.ITextModel,
+  startLine: number,
+  endLine: number,
+  startColumn: number = 1,
+  endColumn?: number
+): monaco.Range {
+  if (endColumn == null) {
+    endColumn = model.getLineMaxColumn(endLine)
```
Comment (CGNonofr) [inline src/tests/utils.ts:21]: nit: you can use it as default value line 18 (:

File: src/tools.ts [hunk 5]
```diff
@@ -201,6 +141,7 @@ function lockCodeUsingDecoration (
         return editorOperations
       }
 
+      editorOperations = computeNewOperationsForLockedCode(model, decorationFilter, editorOperations, withDecoration)
```
Comment (CGNonofr) [inline src/tools.ts:144]: maybe we can find a more meaningful name :thinking: like `tryIgnoreLockedCode` or something?

File: src/tools/utils/editorOperationUtils.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  return splitText.map((text, index) => createNewOperation(oldOperation, editableRanges[index]!, text, index))
+}
+
+function splitOperationText (
+  model: monaco.editor.ITextModel,
+  uneditableRanges: monaco.Range[],
+  text: string | null
+): (string | null)[] {
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:40]: how can it be null?

File: src/tools/utils/editorOperationUtils.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  if (text == null || text === '') {
+    return uneditableRanges.length > 0 ? [] : [text]
+  }
+
+  const splitText: string[] = []
+  const uneditableRangesText = uneditableRanges.map(range => model.getValueInRange(range))
+  let currentRange: number = 0
+  let textToSplit: string = text
+  while (textToSplit !== '' && currentRange < uneditableRangesText.length) {
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:49]: nit: `textToSplit.length > 0` seems clearer to me (we're looping until the string is empty)

also, wouldn't something like `remainingText` be more relevant?

File: src/tools/utils/editorOperationUtils.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  if (editableRanges.length <= 0 || splitText.length <= 0) {
+    return [oldOperation]
+  }
+
+  return splitText.map((text, index) => createNewOperation(oldOperation, editableRanges[index]!, text, index))
+}
+
+function splitOperationText (
+  model: monaco.editor.ITextModel,
+  uneditableRanges: monaco.Range[],
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:38]: `uneditableRangesInOperationRange`?

File: src/tools/utils/editorOperationUtils.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  decorationFilter: (decoration: monaco.editor.IModelDecoration) => boolean,
+  editorOperations: ValidAnnotatedEditOperation[],
+  withDecoration: boolean
+): ValidAnnotatedEditOperation[] {
+  const uneditableRanges = getLockedRanges(model, decorationFilter, withDecoration)
+  if (uneditableRanges.length <= 0) {
+    return editorOperations
+  }
+
+  return splitOperationsForLockedCode(model, editorOperations, uneditableRanges)
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:113]: this function name should probably be similar to the called function name

File: src/tools/utils/editorOperationUtils.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  splitText: (string | null)[]
+): ValidAnnotatedEditOperation[] {
+  if (editableRanges.length <= 0 || splitText.length <= 0) {
+    return [oldOperation]
+  }
+
+  return splitText.map((text, index) => createNewOperation(oldOperation, editableRanges[index]!, text, index))
+}
+
+function splitOperationText (
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:36]: What about `tryIgnoreLockedCodeForOperation`?

File: src/tools/utils/editorOperationUtils.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  const uneditableRangesText = uneditableRanges.map(range => model.getValueInRange(range))
+  let currentRange: number = 0
+  let textToSplit: string = text
+  while (textToSplit !== '' && currentRange < uneditableRangesText.length) {
+    const rangeText = uneditableRangesText[currentRange]
+    if (rangeText != null && rangeText !== '') {
+      const rangeTextIndex = textToSplit.indexOf(rangeText)
+
+      if (rangeTextIndex === -1) {
+        return []
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:55]: What about throwing an error instead?

File: src/tools/utils/editorOperationUtils.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  return splitText.map((text, index) => createNewOperation(oldOperation, editableRanges[index]!, text, index))
+}
+
+function splitOperationText (
+  model: monaco.editor.ITextModel,
+  uneditableRanges: monaco.Range[],
+  text: string | null
+): (string | null)[] {
+  if (text == null || text === '') {
+    return uneditableRanges.length > 0 ? [] : [text]
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:42]: What about throwing an error instead here too?

File: tsconfig.json [hunk 13]
```diff
@@ -19,7 +19,15 @@
...
     "target": "es2020"
   },
-  "include": ["src", "scripts/update-extensions.ts", "rollup.config.ts"],
+  "include": [
+    "src",
+    "scripts/update-extensions.ts",
+    "rollup.config.ts",
+    "jest.config.ts",
+    "src/tests",
+    "jest",
```
Comment (CGNonofr) [inline tsconfig.json:28]: what is "jest"? also It seems src/tests is included in src/ :)

File: src/tests/utils.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+export function createDefaultTestLockedCodeRanges (model: monaco.editor.ITextModel): monaco.Range[] {
+  return [
+    createTestRange(model, 3, 5),
+    createTestRange(model, 12, 14),
+    createTestRange(model, 18, 20)
+  ]
+}
+
+export function canTestOperationsEditRanges (splitOperations: ValidAnnotatedEditOperation[], uneditableRanges: monaco.Range[]): boolean {
```
Comment (CGNonofr) [inline src/tests/utils.ts:50]: I'm confused about the meaning of this function name :thinking:

File: src/tests/editorOperationUtils/lockedCode.test.ts [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    const model = createDefaultTestModel()
+    const uneditableRanges = createDefaultTestLockedCodeRanges(model)
+
+    const operationRange = createTestRange(model, 4, 4)
+    const operation = createTestOperation(operationRange, '// tata')
+    const splitOperations = splitOperationsForLockedCode(model, [operation], uneditableRanges)
+
+    expect(splitOperations.length).toEqual(1)
+    expect(splitOperations[0]).toEqual(createTestOperation(operationRange, '// tata'))
+    expect(canTestOperationsEditRanges(splitOperations, uneditableRanges)).toBe(false)
```
Comment (CGNonofr) [inline src/tests/editorOperationUtils/lockedCode.test.ts:29]: what about applying the edit to the editor and check the final editor content?

File: src/tools.ts [hunk 16]
```diff
@@ -28,6 +17,166 @@ function getRangesFromDecorations (
...
+  if (model == null) {
+    return { startIndex: null, endIndex: null }
+  }
+
+  const editorValue = editor.getValue()
+  const rangeValue = editorValue.slice(model.getOffsetAt(range.getStartPosition()), model.getOffsetAt(range.getEndPosition()))
+  const startIndex = text.indexOf(rangeValue)
+  return {
+    startIndex,
+    endIndex: startIndex + rangeValue.length
```
[the remark this answers is not in the document]
Comment (KaiSaba) [reply src/tools.ts]: This code was in an outdated commit and has been removed with the refactor.

File: src/tests/utils.ts (hunk 4 above)
Comment (KaiSaba) [reply src/tests/utils.ts]: [test: simplify createTestRange function](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/8ec243327bd2e8a240bd3ab1e01f1207a3c3e81d)

File: src/tools.ts (hunk 5 above)
Comment (KaiSaba) [reply src/tools.ts]: [refactor: rename computeNewOperationsForLockedCode to tryIgnoreLockedCode](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/0759eb0e9e1235de24c289da9b02cc7e9a14580d)

File: src/tools/utils/editorOperationUtils.ts (hunk 12 above)
Comment (KaiSaba) [reply src/tools/utils/editorOperationUtils.ts]: [fix: throw custom error when operation is invalid with locked code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/bbf980c6092682d7f7c3cd531813a815c518a63c)

File: src/tools/utils/editorOperationUtils.ts (hunk 11 above)
Comment (KaiSaba) [reply src/tools/utils/editorOperationUtils.ts]: [fix: throw custom error when operation is invalid with locked code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/bbf980c6092682d7f7c3cd531813a815c518a63c)

File: src/tools/utils/editorOperationUtils.ts (hunk 10 above)
Comment (KaiSaba) [reply src/tools/utils/editorOperationUtils.ts]: [refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

File: src/tools/utils/editorOperationUtils.ts (hunk 9 above)
Comment (KaiSaba) [reply src/tools/utils/editorOperationUtils.ts]: [refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

File: src/tools/utils/editorOperationUtils.ts (hunk 7 above)
Comment (KaiSaba) [reply src/tools/utils/editorOperationUtils.ts]: [refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

**Comment by KaiSaba** [comment]
@CGNonofr I added 2 commits:
- [cleanup: remove unneeded console.info](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/7c5a6752f3acff13885667ac95eab65e51807595)
- [fix: removed line bug when pasting code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/b74967419885ed7391c1d5ab39e094f762d81b22)

File: src/tools/utils/editorOperationUtils.ts [hunk 17]
```diff
@@ -2,6 +2,14 @@ import * as monaco from 'monaco-editor'
 import { ValidAnnotatedEditOperation } from 'vscode/vscode/vs/editor/common/model'
 import { getLockedRanges, minusRanges } from './rangeUtils'
 
+export class LockedCodeError extends Error {
+  constructor (message: string) {
+    super(message)
+    this.name = 'LockedCodeError'
+    Object.setPrototypeOf(this, LockedCodeError.prototype)
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts:9]: What is it for?

File: src/tools/utils/editorOperationUtils.ts [hunk 18]
```diff
@@ -26,7 +34,7 @@ function createNewOperationsFromRanges (
   editableRanges: monaco.Range[],
   splitText: (string | null)[]
 ): ValidAnnotatedEditOperation[] {
-  if (editableRanges.length <= 0 || splitText.length <= 0) {
```
Comment (CGNonofr) [inline src/tools/utils/editorOperationUtils.ts]: why removing it?

**Review by CGNonofr — APPROVED** [review]

**Pull request merged**